llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -11874,6 +11874,47 @@ bool TargetLowering::LegalizeSetCCCondCode(SelectionDAG &DAG, EVT VT,
       return true;
     }
 
+    // Special case: expand i1 comparisons using logical operations.
+    if (OpVT == MVT::i1) {
+      SDValue Ret;
+      switch (CCCode) {
+      default:
+        llvm_unreachable("Unknown integer setcc!");
+      case ISD::SETEQ: // X == Y  -->  ~(X ^ Y)
+        Ret = DAG.getNOT(dl, DAG.getNode(ISD::XOR, dl, MVT::i1, LHS, RHS),
```

Comment:
Does this need to be guarded by `TLI.isLegal(ISD::XOR)`?

Reply:
I'd prefer not to. NVPTX is the only target using these expansions and we don't have any way to expand if not. I don't see what the point of adding them would be.

Reply:
My question boils down to: will `xor` be legalized if it's generated for a target that doesn't support it? In other words, can `Expand` generate illegal code? 

My assumption was no, but based on your answer, I'm gathering that `Expand` can generate illegal code and rely on the legalizer to legalize it. Is that correct?

Reply:
I think the issue here is that there's more than one possible interpretation of "lower SetCC(i1) differently". NVPTX can do it via logical ops on predicates, other targets may not. On other targets it may be something else. Whoever would use "setCondCodeAction(...,Expand)" would have no idea what it would expand into. This should at least be documented. If/when we would encounter a target which would have a different way of expanding setcc, we would probably need a more nuanced hint to tell LLVM which flavor we actually want.

I think at the moment only Hexagon uses `setCondCodeAction(...,i1, Expand)`. It would be good to check what effect this change has on them.